service/route_handlers.go

```diff
@@ -554,25 +635,83 @@ func HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml
 		Path:   "/",
 	}
 	http.SetCookie(w, tokenCookie)
-
-	log.Debugf("redirecting the user with token to %v", redirectURL)
+	log.Debugf("redirecting the user with cookie %v to %v", tokenCookie, redirectURL)
 
 	http.Redirect(w, r, redirectURL, http.StatusFound)
 	return
 }
 
+func PostSamlTokenHTML(w http.ResponseWriter, r *http.Request) {
+	var token, finalRedirectURL string
+	var secure bool
+	var err error
+
+	r.ParseForm()
+
+	if len(r.Form["token"]) > 0 { //array of string
+		token = r.Form["token"][0]
+	}
+
+	if len(r.Form["finalRedirectURL"]) > 0 {
+		finalRedirectURL = r.Form["finalRedirectURL"][0]
+	}
+
+	if len(r.Form["secure"]) > 0 {
+		secure, err = strconv.ParseBool(r.Form["secure"][0])
+		if err != nil {
+			log.Errorf("PostSamlTokenHTML: Error in parsing secure flag for token: %v", err)
+			return
+		}
+	}
+
+	tokenCookie := &http.Cookie{
+		Name:   "token",
+		Value:  token,
+		Secure: secure,
+		MaxAge: 0,
+		Path:   "/",
+	}
+	http.SetCookie(w, tokenCookie)
+	log.Debugf("redirecting the user with cookie %v to %v", tokenCookie, finalRedirectURL)
+
+	http.Redirect(w, r, finalRedirectURL, http.StatusFound)
+}
+
 func isWhitelisted(redirectBackBaseValue string, redirectWhitelist string) bool {
+	var redirectBackBaseDomainNonPort string
 	if redirectBackBaseValue == server.GetRancherAPIHost() {
 		return true
 	}
+	redirectURL, err := url.Parse(redirectBackBaseValue)
+	if err != nil {
+		log.Errorf("Error in parsing redirect URL")
+		return false
+	}
+	redirectBackBaseDomain := redirectURL.Host
+	if strings.Contains(redirectBackBaseDomain, ":") {
+		d := strings.SplitN(redirectBackBaseDomain, ":", 2)
+		redirectBackBaseDomainNonPort = d[0]
+	}
 
 	// comma separated list of whitelisted domains to redirect to
 	if redirectWhitelist != "" {
 		whitelistedDomains := strings.Split(redirectWhitelist, ",")
+		log.Debugf("Whitelisted domains provided: %v, redirectBackBase: %v, redirectBackBaseNonPort: %v", whitelistedDomains,
+			redirectBackBaseDomain, redirectBackBaseDomainNonPort)
 		for _, w := range whitelistedDomains {
-			if redirectBackBaseValue == w {
+			if w == redirectBackBaseDomain || w == redirectBackBaseDomainNonPort {
+				log.Debugf("Whitelisted domain matched: %v", w)
 				return true
 			}
+
+			if strings.Contains(w, ":") {
+				d := strings.SplitN(w, ":", 2)
+				log.Debugf("Domain: %v", d[0])
+				if d[0] == redirectBackBaseDomain || d[0] == redirectBackBaseDomainNonPort {
+					log.Debugf("Whitelisted domain matched: %v", d[0])
+					return true
+				}
+			}
```

Comment:
Since we are expecting that Admin can enter the whitelisted string as just a domain x.com OR complete url x.com:<port> - we should check if the whitelisted value has port and make sure it matches the port presented by 'redirectBackBaseDomain'.

If whitelisted value is just a domain, we can check only the domain matches irrespective of the port

Reply:
Basically we need to ensure the whitelisted value matches as-is.